experimental/LieAlgebras/src/PathModel.jl

```diff
@@ -0,0 +1,617 @@
+
+function falpha!(b::AbstractCrystalElem, i::Int, n::Int)
+  for _ in 1:n
+    falpha!(b, i)
+  end
+  return b
+end
+
+function falpha!(b::AbstractCrystalElem, i::Vector{Int}, n::Vector{Int})
+  @req length(i) == length(n) "i and n must have same length"
+  
+  for l in length(i):-1:1
+    falpha!(b, i[l], n[l])
+  end
+  return b
+end
+
+# other
+
+function adapted_string(p::AbstractCrystalElem, rdec::Vector{Int})
+  s = zero(rdec)
+  b = deepcopy(p)
+  for i in 1:length(rdec)
+    s[i] = Base.eps(b, rdec[i]) # TODO: this will not work for non normal crystals
+    ealpha!(b, rdec[i], s[i])
+  end
+  return s
+end
+
+# LS Path Model 
+
+struct LSPathModel
+  wt::WeightLatticeElem
+  ext::Dict{Vector{UInt8}, WeightLatticeElem} # map from Weyl group elements to extremal weights
+  
+  function LSPathModel(wt::WeightLatticeElem)
+    @req is_dominant(wt) "weight must be dominant"
+    return new(wt, Dict(one(weyl_group(root_system(wt))) => wt))
+  end
+end
+
+# Young Tableu Path Model
+
+struct YTPathModel <: AbstractCrystal
+  wt::WeightLatticeElem
+end
+
+struct YTPathModelElem <: AbstractCrystalElem
+  parent::YTPathModel
+  t::YoungTableau
+end
+
+function Base.show(io::IO, mime::MIME"text/plain", P::LSPathModel)
+  #@show_name(io, P)
+  #@show_special(io, mime, P)
+  #io = pretty(io)
+  print(io, "LS path model for ")
+  show(io, mime, P.wt)
+end
+
+function Base.show(io::IO, P::LSPathModel)
+  #@show_name(io, R)
+  #@show_special(io, R)
+  if is_terse(io)
+    print(io, "LS path model")
+  else
+    print(io, "LS path model for $(P.wt)")
+  end
+end
+
+@doc raw"""
+    ls_path_model(wt::WeightLatticeElem) -> LSPathModel
+
+Return the LS path model for the dominant weight `wt`.
+"""
+function ls_path_model(wt::WeightLatticeElem)
+  return LSPathModel(wt)
+end
+
+@doc raw"""
+    ls_path_model(R::RootSystem, wt::Vector{<:IntegerUnion}) -> LSPathModel
+
+Return the LS path model for the root system `R` and dominant weight `wt`.
+"""
+function ls_path_model(R::RootSystem, wt::Vector{<:IntegerUnion})
+  return ls_path_model(WeightLatticeElem(R, wt))
+end
+
+function _extremal_weight(P::LSPathModel, w::WeylGroupElem)
+  wt = get(P.ext, word(w), nothing)
+  if isnothing(wt)
+    # we need to make a copy, because w may be modified
+    return get!(P.ext, deepcopy(word(w)), w*P.wt)
+  end
+  
+  return wt
+end
+
+function (P::LSPathModel)(wv::Vector{WeylGroupElem}, tv::Vector{<:RationalUnion})
+  return LSPathModelElem(P, [LSPathSegment(t, w) for (t, w) in Iterators.zip(tv, wv)])
+end
+
+function (P::LSPathModel)(p::YTPathModelElem)
+  @req P.wt == parent(p).wt "dominant weights must match"
+  
+  pp = dominant_path(P)
+  i = Int.(word(longest_element(weyl_group(P.wt))))
+  falpha!(pp, i, adapted_string(p, i))
+  return pp
+end
+
+function root_system(P::LSPathModel)
+  return root_system(P.wt)
+end
+
+# LSPathModelElem
+
+struct LSPathSegment
+  t::QQFieldElem
+  w::WeylGroupElem
+end
+
+function Base.:(==)(s1::LSPathSegment, s2::LSPathSegment)
```

Comment:
When defining a custom `==` method, make sure to also (two argument) `hash` method if there is even a tiny chance of anyone ever trying to use these as keys in a dictionary or apply `unique` to them etc.

Seems you have one for `LSPathModelElem`, but not for `LSPathSegment`; and the former ends up hashing elements of the latter type, so I guess you really need this.